### PR TITLE
Add damped Newton's option to the 2-phase KKS nested solution method

### PR DESCRIPTION
**Pull request by xueyang94**
## Reason
In the KKS nested solution implementation (#19981), only the direct Newton's method is available for solving phase concentration variables locally as material properties. When dealing with highly nonlinear phase free energies, e.g, logarithmic functions, the concentration updates in the nested Newton may be too large that push the concentrations far away from the roots. Here, we add the option of damped Newton's method, which multiplies the update by an input factor that is between zero and one, therefore, shrinks the update of concentrations.

## Design

In KKSPhaseConcentrationMaterial class, three options are offered regarding damping: no damping at all, damping for once, and damping until the concentrations are between the lower and upper bounds provided by the user. The damping implementation is in NestedSolve class.

## Impact

This enhancement adds new features to the 2-phase KKS nested method.

@dschwen @jiangwen84  @tonkmr

### CODE REVIEW

**Comment by xueyang94** [comment]
@chaitanyaBhave26

File: framework/include/utils/NestedSolve.h [hunk 1]
```diff
@@ -355,3 +367,123 @@ NestedSolve::nonlinear(V & guess, T compute)
...
+template <typename V, typename T, typename B>
+void
+NestedSolve::nonlinear(V & guess,
+                       T compute,
+                       Real _damping_factor,
+                       B ci_lower_bounds,
+                       B ci_upper_bounds,
+                       unsigned int _num_eta,
+                       unsigned int _num_c,
+                       std::string _damped_newton)
```
Comment (dschwen) [inline framework/include/utils/NestedSolve.h:407]: this should be an `enum class` and drop teh leading underscore (those are reserved for member variables)

File: framework/include/utils/NestedSolve.h [hunk 2]
```diff
@@ -355,3 +367,123 @@ NestedSolve::nonlinear(V & guess, T compute)
...
+    if (_sum_independent > 1 || _sum_independent == 1 || _sum_independent < 0 ||
+        _sum_independent == 0)
+      return false;
+  }
+  return true;
+}
+
+template <typename V, typename T, typename B>
+void
+NestedSolve::nonlinear(V & guess,
```
Comment (dschwen) [inline framework/include/utils/NestedSolve.h:400]: it should be possible to not make this a separate function. I see a bit of code duplication here.

**Comment by dschwen** [comment]
I have a PR that adds a bounded solve and which significantly rewrites the NestedSolve class: https://github.com/idaholab/moose/pull/20409 
This will be a bit of a merge conflict nightmare either way :-(

File: framework/include/utils/NestedSolve.h (hunk 1 above)
Comment (xueyang94) [reply framework/include/utils/NestedSolve.h]: I changed _damped_newton to enum class DampedNewton.

File: framework/include/utils/NestedSolve.h (hunk 2 above)
Comment (xueyang94) [reply framework/include/utils/NestedSolve.h:375]: I made condition() to be a lambda function inside NestedSolve::nonlinear().

**Comment by xueyang94** [comment]
> I have a PR that adds a bounded solve and which significantly rewrites the NestedSolve class: #20409 This will be a bit of a merge conflict nightmare either way :-(

If you would merge yours first, I can add the damping capability to the new version of NestedSolve.h.

**Comment by tophmatthews** [comment]
+1 to this...would like to use dampening in `NestedSolve`

**Comment by xueyang94** [comment]
> +1 to this...would like to use dampening in `NestedSolve`

Now that the nested approach for the 2-phase KKS model has been merged (#19981), the damping feature should be ready to add @chaitanyaBhave26

**Comment by dschwen** [comment]
Almost, there is the ad nested solve PR we need to get in first.

**Comment by dschwen** [comment]
We're still working on integrating this

**Comment by tophmatthews** [comment]
I'm very interested in the damping part. Can that be split off from the application to `KKSPhaseConcentrationMaterial`?

**Pull request closed without merging**